### PR TITLE
Global parameter to force exp(x) to be Pow(E, x)

### PR DESCRIPTION
**Pull request by Upabjojr**
Alternative to https://github.com/sympy/sympy/pull/20819
Refactory of #13211, splitting the previous PR into simple steps.

#### Release Notes

NO ENTRY

- [ ] how is the (ascii, not unicode) pretty printer supposed to print `E`? Currently, in SymPy's master, it`s `e` when inside `exp( )`, otherwise `E`. This is inconsistent.
- [ ] detect duplicate code and pack it into functions.

### CODE REVIEW

**Comment by Upabjojr** [comment]
Original issue: https://github.com/sympy/sympy/issues/4898

**Comment by asmeurer** [comment]
I still don't think this should be part of any public API (so no release notes mentions of it). Ideally we would do the whole migration before any of it gets into a release, even if that is split into multiple PRs. 

>  how is the (ascii, not unicode) pretty printer supposed to print E? Currently, in SymPy's master, its ewhen insideexp( ), otherwise E`. This is inconsistent.

I'd say that's a bug. The pretty printer should pretty print things using mathematical notation, even the ASCII pretty printer, meaning e should pretty print as `e`. Only the str printer should print `E`, because that is what the Python variable is called and str should be valid Python code to create the expression.

**Comment by Upabjojr** [comment]
> I still don't think this should be part of any public API (so no release notes mentions of it).

This only element added is the `_exp_is_pow` context manager, whose name starts with an underscore and is therefore private.

Context manager/global parameter approach is actually quite, as you can switch back and forth between `exp` and `Pow` as you wish.

This PR could either be merged and the work continued on other PRs, or extended. I'm still pretty sure we won't have everything ready in just one PR.

**Comment by Upabjojr** [comment]
Is there a PyTest decorator to run the tests twice in two different context managers?

**Comment by asmeurer** [comment]
It would be simpler to make an environment variable (but again, this should be temporary and NOT public API).

**Comment by Upabjojr** [comment]
> It would be simpler to make an environment variable (but again, this should be temporary and NOT public API).

Well, that doesn't really seem simpler to me.

**Comment by oscarbenjamin** [comment]
Something like an environment variable would be simpler if the goal is to run the whole or a large part of the test suite under different settings. It would mean that you wouldn't need to edit much of the test code: you could just run the tests under both settings of the environment variable.

**Comment by Upabjojr** [comment]
> Something like an environment variable would be simpler if the goal is to run the whole or a large part of the test suite under different settings. It would mean that you wouldn't need to edit much of the test code: you could just run the tests under both settings of the environment variable.

I'm going for a decorator on the tests. An environment variable doesn't really make sense, as the transformation will be split over many PRs. So a decorator to run the single test twice is optimal to keep track of which tests have already been ported to the new operational mode.

**Comment by Upabjojr** [comment]
So, any objection to merging this PR?

I have been careful to keep this new behaviour hidden (users will not notice it unless they inspect the code). SymPy's behaviour does not change. A decorator can be added to the tests in order to test both `exp(x)` and `Pow(E, x)` modes.

**Comment by asmeurer** [comment]
Is there some way in this PR to see what tests fail with the Pow-exp mode?

> I have been careful to keep this new behaviour hidden (users will not notice it unless they inspect the code). SymPy's behaviour does not change. A decorator can be added to the tests in order to test both exp(x) and Pow(E, x) modes.

As I noted the release notes entries should be removed too. We can add a final entry for the full transition once the last PR is done.

File: sympy/testing/pytest.py [hunk 1]
```diff
@@ -214,6 +214,19 @@ def warns(warningcls, *, match=''):
             raise Failed(msg)
 
 
+def _both_exp_pow(func):
```
Comment (asmeurer) [inline sympy/testing/pytest.py:217]: Can you add a comment here explaining why this is here and that it is only temporary while this fix is being done.

**Comment by asmeurer** [comment]
> I'm going for a decorator on the tests. An environment variable doesn't really make sense, as the transformation will be split over many PRs. So a decorator to run the single test twice is optimal to keep track of which tests have already been ported to the new operational mode.

Hopefully there aren't any issues relating to modifying the global context. These would most likely be due to the cache. What I mean is an issue where running a test once in either context (Pow mode or exp mode) is fine, but the test fails if you run it once in one and then again in another. Such issues would be very difficult to debug, and would ultimately be irrelevant to the final implementation, since we don't _actually_ need the cache (or whatever) to support both modes. The cache could also potentially cause a test to pass when it really should fail with this method, if the second run doesn't actually compute something new.

**Comment by asmeurer** [comment]
Most of the stuff that has been added to Pow does not need to be special-cased for Exp1 (that's the whole point of merging them). Since `x**y` = `exp(log(x)*y)` anything formula for base e should also have a corresponding formula for base b. Is the plan is to fix those in a future PR? It would be helpful if you could outline your plan of how you intend to fix this.

**Comment by oscarbenjamin** [comment]
This is using the same mechanism as the other global parameters so it should clear the cache when the parameter is changed. There is still the possibility for inconsistent results when changing global parameters in a multithreaded context though.

**Comment by Upabjojr** [comment]
> Is there some way in this PR to see what tests fail with the Pow-exp mode?

All the Risch- and Meijer G algorithms are going to fail if we switch to `Pow(E, ...)` there. The new-style assumptions as well. In short... a high percentage of SymPy's tests will fail.

> The cache could also potentially cause a test to pass when it really should fail with this method, if the second run doesn't actually compute something new.

OK, so maybe the `_both_exp_pow` decorator should clear the cache.


> Most of the stuff that has been added to Pow does not need to be special-cased for Exp1 (that's the whole point of merging them).

I would prefer to wait until all unit tests work with `Pow(E, ...)` before starting to simplify the code.

**Comment by oscarbenjamin** [comment]
> It would be helpful if you could outline your plan of how you intend to fix this.

I agree with @asmeurer here. It would be good to have a document explaining the plan. It could go on the wiki for example or perhaps the new sympeps repo. I don't think it needs to be fully detailed but it should explain:

1. How the work will be completed incrementally
2. What the end result will be
3. How the final breaking change would work
4. What users or downstream library authors should do to make their code work across this transition

**Comment by asmeurer** [comment]
Something to consider: if we want to try to maintain backwards compatibility with deprecation as much as possible, and think we should, as discussed at https://github.com/sympy/sympy/pull/20819, then we should implement that first. That will allow us to transition SymPy itself in a backwards compatible way, except for the deprecation warnings, which can be removed one-by-one. And a side effect is that we can be more confident that we really are maintaining backwards compatibility (with warnings) as much as possible if we can make the SymPy test suite pass with the `exp` shim.

**Comment by Upabjojr** [comment]
> * How the work will be completed incrementally

* Finish the compatibility `Pow(E, x)` (e.g. assumptions are probably not working yet, check expression rewriting, etc...)
* Detect all usages certain expressions and replace them properly:
  * `exp(x).args[0]` ==> `exp(x).exp`,
  * `expr.has(exp)` ==> `expr.has_exp()`,
  * `isinstance(expr, exp)` ==> instancecheck trick in `exp`,

> * What the end result will be

The target as of now is to make both `exp(x)` and `Pow(E, x)` work seamlessly at the same time. Users will be able to choose which one to use. No plans on deprecation yet. There will be a discussion later.

> * How the final breaking change would work

No plans on breaking anything yet. We can open a discussion once both `Pow(E, x)` and `exp(x)` pass all tests.

> * What users or downstream library authors should do to make their code work across this transition

This will be decided if we decide to deprecate `exp(x)`. We will open a new discussion as soon as both expressions work.

Though, I imagine we will simply add a deprecation warning on `exp(x).args[0]` telling users to use `exp(x).exp` instead. Same for `expr.has(exp)`, `isinstance(expr, exp)`, and `expr.rewrite(exp)`. In reality, only `exp(x).args[0]` and `exp(x) is exp` are probably the only breaking parts of code.

**Comment by asmeurer** [comment]
> exp(x).args[0] ==> exp(x).exp,

Also a good idea to fix this for Pow expressions if you come across it. Unless the code is unpacking args like `base, exp = pow.args`, `pow.base` and `pow.exp` are better than `pow.args[0]` and `pow.args[1]`.

> expr.has(exp) ==> expr.has_exp(),

I didn't test it, but looking at the code for `_has`, I think the isinstance trick would make `has(exp)` work. That would be preferable to a new method just for this. 

> The target as of now is to make both exp(x) and Pow(E, x) work seamlessly at the same time. Users will be able to choose which one to use. No plans on deprecation yet. There will be a discussion later.

What's the motivation for this? Wouldn't this make the code more complicated, as any code that handles one also needs to handle the other? Whereas if we just used Pow(E, x) we only need to make sure that code that previously handled `exp` now handles it in the Pow form. 

> Though, I imagine we will simply add a deprecation warning on exp(x).args[0] telling users to use exp(x).exp instead. Same for expr.has(exp), isinstance(expr, exp), and expr.rewrite(exp). In reality, only exp(x).args[0] and exp(x) is exp are probably the only breaking parts of code.

I hadn't considered `exp(x).args`. That is indeed problematic. Even if we make `isinstance(exp(x), exp)` work, code that does something like

```py
if isinstance(expr, exp):
    x = expr.args[0]
```

would be wrong. There's really no way to deprecate it, as far as I can see. We shouldn't make `Pow(E, x).args` give `(x,)`. Probably we just have to make the change here and make a special note of it in the release notes.

**Comment by Upabjojr** [comment]
> What's the motivation for this?

First make __all__ unit tests work with both `Pow(E, x)` and `exp(x)`, then start deprecating `exp(x)`.



> I hadn't considered `exp(x).args`.

```python

class exp( ... ):
    ...

    @property
    def args(self):
        SymPyDeprecationWarning("exp( ).args[0] is deprecated, use exp( ).exp").warn()
        return self._args
```

**Comment by asmeurer** [comment]
The problem is when exp(x) becomes Pow(E, x) we can't prevent Pow(E, x).args[0] (Pow.args shouldn't be deprecated).

**Comment by Upabjojr** [comment]
> The problem is when exp(x) becomes Pow(E, x) we can't prevent Pow(E, x).args[0] (Pow.args shouldn't be deprecated).

There should be a period in which users are warned to remove all calls to `exp( ).args`. Currently, `.exp` works on both `Pow` and `exp`, so that should be the way to access the argument of `exp`.

**Comment by Upabjojr** [comment]
Or maybe... a two-step deprecation:

1. At first, warn users not to use `exp(x).args[0]`, and maybe add support for `exp(x).args[1]` as an alias of `exp(x).args[0]`.
2. Second phase, `exp(x).args` should throw an exception if called on `exp(x).args[0]`. Somehow this should be fixed on the internal methods (in order not to break stuff like replacements and substitutions).
3. Finally, replace `exp(x)` with `Pow(E, x)`.

**Comment by Upabjojr** [comment]
So, any other objections to merging this PR?

**Comment by asmeurer** [comment]
I think what is in this PR is fine. 

I'd like to continue to discuss the process for the rest, though. Is there an issue we can move it to? 

I'm not convinced that doing a multistage deprecation is worth it. I'm curious what other's thoughts are. For one thing, we can't deprecate `exp(x).args`. There is a lot of code that uses `.args` on expressions generically (they don't know what the type of the expression is), and we don't want that to stop working for exponentials. Also, I would personally rather see this all done in one go. Your suggested process would require at least one intermediate release where we have this mixed behavior. Actually you have two phases so that means two releases. So it could be well over a year before this gets completed. But if you are willing to do the work now to finish this, then I would like to see this done now.

**Comment by Upabjojr** [comment]
Let me know... this PR could potentially be merged. Everything is hidden and end users won't notice anything.

**Comment by oscarbenjamin** [comment]
> I'm not convinced that doing a multistage deprecation is worth it. I'm curious what other's thoughts are.

I don't think this is a multistage deprecation. It's a multistage implementation of a parallel but not used by default implementation. In this PR nothing is actually deprecated and nothing is enabled by default (if I've understood correctly).

It's similar to `with distribute(False)` which gives the behaviour we ideally want but can't actually be used because it breaks everything. If there is a sustained effort over time to make the codebase work under both modes then it could become usable and we could consider at some point in the future flipping the default behaviour. The final change would have to be a breaking change in a breaking release though: it isn't actually possible to have the new behaviour at the same time as allowing the old behaviour but emitting a deprecation warning.

**Comment by Upabjojr** [comment]
So, shall we proceed?

**Comment by asmeurer** [comment]
I think this sounds OK to me. Are you OK with it @oscarbenjamin?

I think the args breaking is something we can't avoid. Maybe we can just put a note in the release notes that it will change in a future version. 

How long do you expect the full changes will take?

**Comment by oscarbenjamin** [comment]
I think this is fine.

It does imply a breaking release at some point in the future. I think we should plan for a sympy 2.0 and think about managing compatibility breaks in a broader way.

**Comment by Upabjojr** [comment]
> How long do you expect the full changes will take?

Hopefully not long. It will take some time to fix all of SymPy's code (especially inside the _integrals_ module).

**Pull request merged**